### PR TITLE
Add infrastructure for databricks

### PR DESCRIPTION
**Pull request by johevemi**
## References

* #91

### CODE REVIEW

File: .github/actions/publish-release-contents/action.yml [hunk 1]
```diff
@@ -31,8 +31,24 @@ runs:
...
 
+    - name: Publish Streaming cluster Infrastructure to release folder
+      shell: bash
+      run: |
+        cp -fR ${{ inputs.REPO_FOLDER_PATH }}/build/databricks_streaming_cluster ${{ inputs.RELEASE_FOLDER_PATH }}/infrastructure/databricks_streaming_cluster 2>/dev/null || :
+
+    - name: Publish streaming-jobs to release folder
+      shell: bash
+      run: |
+        cp -fR ${{ inputs.REPO_FOLDER_PATH }}/source/databricks/streaming-jobs ${{ inputs.RELEASE_FOLDER_PATH }} 2>/dev/null || :
```
Comment (BjarkeMeier) [inline .github/actions/publish-release-contents/action.yml:42]:
```suggestion
    - name: Publish Databricks jobs to release folder
      shell: bash
      run: |
        cp -fR ${{ inputs.REPO_FOLDER_PATH }}/source/databricks/jobs ${{ inputs.RELEASE_FOLDER_PATH }} 2>/dev/null || :
```

File: .github/actions/publish-release-contents/action.yml [hunk 2]
```diff
@@ -31,8 +31,24 @@ runs:
         mkdir ${{ inputs.RELEASE_FOLDER_PATH }}/infrastructure
         cp -fR ${{ inputs.REPO_FOLDER_PATH }}/build/infrastructure/main ${{ inputs.RELEASE_FOLDER_PATH }}/infrastructure/main 2>/dev/null || :
 
+    - name: Publish Streaming cluster Infrastructure to release folder
+      shell: bash
+      run: |
+        cp -fR ${{ inputs.REPO_FOLDER_PATH }}/build/databricks_streaming_cluster ${{ inputs.RELEASE_FOLDER_PATH }}/infrastructure/databricks_streaming_cluster 2>/dev/null || :
```
Comment (BjarkeMeier) [inline .github/actions/publish-release-contents/action.yml:37]:
```suggestion
    - name: Publish Databricks cluster terraform to release folder
      shell: bash
      run: |
        cp -fR ${{ inputs.REPO_FOLDER_PATH }}/build/databricks_cluster ${{ inputs.RELEASE_FOLDER_PATH }}/infrastructure/databricks_cluster 2>/dev/null || :
```

File: .github/actions/publish-release-contents/action.yml [hunk 3]
```diff
@@ -31,8 +31,24 @@ runs:
...
+
+    - name: Publish streaming-jobs to release folder
+      shell: bash
+      run: |
+        cp -fR ${{ inputs.REPO_FOLDER_PATH }}/source/databricks/streaming-jobs ${{ inputs.RELEASE_FOLDER_PATH }} 2>/dev/null || :
+
+    - name: Publish wheel to release folder
```
Comment (BjarkeMeier) [inline .github/actions/publish-release-contents/action.yml:44]:
```suggestion
    - name: Publish Python wheel to release folder
```

File: .github/workflows/cd.yml [hunk 4]
```diff
@@ -46,3 +46,10 @@ jobs:
       AZURE_SPN_ID: ${{ secrets.AZURE_SPN_ID_OIDC }}
       AZURE_KEYVAULT_URL: ${{ secrets.AZURE_KEYVAULT_URL }}
       AZURE_SECRETS_KEYVAULT_URL: ${{ secrets.AZURE_SECRETS_KEYVAULT_URL }}
+
+  update_coverage_report_python:
```
Comment (BjarkeMeier) [inline .github/workflows/cd.yml:50]:
```suggestion
  update_python_coverage_report:
```

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -49,6 +43,28 @@ jobs:
...
+  terraform_validate:
+    uses: Energinet-DataHub/.github/.github/workflows/terraform-validate.yml@7.6.0
+    with:
+      TERRAFORM_WORKING_DIR_PATH: './build/infrastructure/main'
+      TERRAFORM_VERSION: '1.2.2'
+  
+  terraform_validate_streaming_cluster:
+    uses: Energinet-DataHub/.github/.github/workflows/terraform-validate.yml@7.6.0
+    with:
+      TERRAFORM_WORKING_DIR_PATH: './build/databricks_streaming_cluster'
```
Comment (BjarkeMeier) [inline .github/workflows/ci.yml:55]:
```suggestion
      TERRAFORM_WORKING_DIR_PATH: './build/databricks_jobs'
```

File: build/databricks_streaming_cluster/main/main.tf [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+data "databricks_spark_version" "latest_lts" {
+  long_term_support = true
+}
+
+resource "databricks_job" "integration_events_persister_streaming_job" {
+  name = "IntegrationEventsPersisterStreamingJob"
+  max_retries = 2
```
Comment (BjarkeMeier) [inline build/databricks_streaming_cluster/main/main.tf:21]: Why not infinite?

File: build/databricks_streaming_cluster/main/main.tf [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    pypi {
+      package = "configargparse==1.2.3"
+    }
+  }
+
+  library {
+    pypi {
+      package = "azure-storage-blob==12.7.1"
+    }
+  }
```
Comment (BjarkeMeier) [inline build/databricks_streaming_cluster/main/main.tf:47]: Use requirements.txt?

File: build/databricks_streaming_cluster/main/main.tf [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+         "--event-hub-connectionstring=${var.evh_wholesale_listen_connection_string}",
+         "--integration-events-path=abfss://integration-events@${data.azurerm_key_vault_secret.st_data_lake_name.value}.dfs.core.windows.net/events",
+         "--integration-events-checkpoint-path=abfss://integration-events@${data.azurerm_key_vault_secret.st_data_lake_name.value}.dfs.core.windows.net/events-checkpoint",
+    ]
+  }
+
+
+  email_notifications {
+    no_alert_for_skipped_runs = true
+  }
```
Comment (BjarkeMeier) [inline build/databricks_streaming_cluster/main/main.tf:67]: Can this be replaced by a hook to the primary action group?

File: build/infrastructure/main/outputs.tf [hunk 9]
```diff
@@ -12,6 +12,18 @@
...
+  value       = module.evh_masterdataevents.primary_connection_strings["listen"]
+  sensitive   = true
+}
+
+output databricks_workspace_url {
+  description = "URL of the created Databricks workspace"
+  value       = data.azurerm_key_vault_secret.dbw_shared_workspace_url.value
+  sensitive   = true
+}
+
```
Comment (BjarkeMeier) [inline build/infrastructure/main/outputs.tf:26]: Why is this output needed?

File: source/databricks/package/integration_events_persister.py [hunk 10]
```diff
@@ -17,17 +17,19 @@
...
 from pyspark.sql import DataFrame
 from pyspark.sql.functions import (
-    year, month, dayofmonth, col)
+    year, month, dayofmonth, col, from_json)
+from package.schemas import (
+    eventhub_integration_events_schema as schema)
 
 
 def integration_events_persister(streamingDf: DataFrame, checkpoint_path: str, integration_events_path: str):
+    streamingDf = streamingDf.withColumn("body", col("body").cast("string")).withColumn("body", from_json(col("body"), schema))
```
Comment (BjarkeMeier) [inline source/databricks/package/integration_events_persister.py:26]: Better readability:
```suggestion
from package.schemas import (
    eventhub_integration_events_schema as schema)


def integration_events_persister(streamingDf: DataFrame, checkpoint_path: str, integration_events_path: str):
    streamingDf = streamingDf
        .withColumn("body", col("body").cast("string"))
        .withColumn("body", from_json(col("body"), schema))
```

File: build/databricks_streaming_cluster/main/providers.tf [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+terraform {
+  required_version = "1.2.2"
+
+  required_providers {
+    databricks = {
```
Comment (BjarkeMeier) [inline build/databricks_streaming_cluster/main/providers.tf:18]: Two weeks ago Databricks [published a GA provider](https://databricks.com/blog/2022/06/22/databricks-terraform-provider-is-now-generally-available.html).

File: build/infrastructure/main/outputs.tf (hunk 9 above)
Comment (defectiveAi) [reply build/infrastructure/main/outputs.tf:26]: I do not think it is.

Comment (lasrinnil) [reply build/infrastructure/main/outputs.tf]: removed

File: build/databricks_streaming_cluster/main/providers.tf (hunk 11 above)
Comment (lasrinnil) [reply build/databricks_streaming_cluster/main/providers.tf:18]: added to list of91

File: .github/actions/publish-release-contents/action.yml (hunk 2 above)
Comment (lasrinnil) [reply .github/actions/publish-release-contents/action.yml:37]: Added to story #91

**Review by defectiveAi — APPROVED** [review]

**Pull request merged**